lib/services/unified_resource.go

```diff
@@ -0,0 +1,675 @@
+	defer c.mu.Unlock()
+	i, found := c.tree.Get(&btreeItem{Item: expected})
+	if !found {
+		return trace.CompareFailed("key %q is not found", string(expected.Key))
+	}
+	existingItem := i.Item
+	if existingItem.Value != expected.Value {
+		return trace.CompareFailed("current value does not match expected for %v", string(expected.Key))
+	}
+	event := Event{
+		Type: types.OpPut,
+		Item: replaceWith,
+	}
+	c.processEvent(event)
+	return nil
+}
+
+type getResult struct {
+	Items []Item
+}
+
+func (c *unifiedResourceCache) getRange(ctx context.Context, startKey, endKey []byte, limit int) getResult {
+	var res getResult
+	c.tree.AscendRange(&btreeItem{Item: Item{Key: startKey}}, &btreeItem{Item: Item{Key: endKey}}, func(item *btreeItem) bool {
+		res.Items = append(res.Items, item.Item)
+		if limit > 0 && len(res.Items) >= limit {
+			return false
+		}
+		return true
+	})
+	return res
+}
+
+func (c *unifiedResourceCache) processEvent(event Event) {
+	switch event.Type {
+	case types.OpPut:
+		item := &btreeItem{Item: event.Item, index: -1}
+		c.tree.ReplaceOrInsert(item)
+	case types.OpDelete:
+		item, found := c.tree.Get(&btreeItem{Item: event.Item})
+		if !found {
+			return
+		}
+		c.tree.Delete(item)
+	default:
+		// skip unsupported record
+	}
+}
+
+// UnifiedResourceWatcherConfig is a UnifiedResourceWatcher configuration.
+type UnifiedResourceWatcherConfig struct {
+	ResourceWatcherConfig
+	NodesGetter
+	DatabaseServersGetter
+	AppServersGetter
+	WindowsDesktopGetter
+	KubernetesClusterGetter
+	SAMLIdpServiceProviderGetter
+}
+
+// UnifiedResourceWatcher is built on top of resourceWatcher to monitor additions
+// and deletions to resources displayable in the UI such as Nodes, Dbs, Apps, etc.
+type UnifiedResourceWatcher struct {
+	*resourceWatcher
+	*unifiedResourceCollector
+}
+
+// Close closes the underlying resource watcher
+func (u *UnifiedResourceWatcher) Close() error {
+	u.resourceWatcher.Close()
+	return u.unifiedResourceCollector.Close()
+}
+
+// GetUnifiedResources returns a list of all resources stored in the current unifiedResourceCollector tree
+func (u *UnifiedResourceWatcher) GetUnifiedResources(ctx context.Context) ([]types.ResourceWithLabels, error) {
+	var resources []types.ResourceWithLabels
+
+	// if the watcher is not initialized or stale, instead of returning nothing, return upstream nodes
+	if !u.IsInitialized() || u.stale {
+		nodes, err := u.NodesGetter.GetNodes(ctx, apidefaults.Namespace)
+		if err != nil {
+			return nil, trace.Wrap(err, "getting nodes while unified resource cache is uninitialized or stale")
+		}
+
+		for _, node := range nodes {
+			resources = append(resources, node)
+		}
+		return resources, nil
+	}
+
+	result, err := u.current.GetRange(ctx, backend.Key(prefix), backend.RangeEnd(backend.Key(prefix)), backend.NoLimit)
+	if err != nil {
+		return nil, trace.Wrap(err, "getting unified resource range")
+	}
+	for _, item := range result.Items {
+		resources = append(resources, item.Value)
+	}
```

Comment:
This is a risky default.  Any resource returned to the caller should be deep copied to prevent concurrent memory modification. If we discover a code path where we are 100% certain no modification occurs we could maybe add some optimizations down the line, but the default should be to clone on get.

Additionally, we should only be returning/copying those resources that match our query/filter (see the node watcher for an example of how to do this, a simple `match fn(Resource) bool` is probably the easiest).